devlib/utils/ssh.py

```
@@ -299,6 +312,469 @@ def __init__(self,
+        written out to the host system using writefile, prior to being moved to
+        the destination on the host.
+        """
+        filename = os.path.basename(source)
+
+        logger.debug("pull_file {} {}".format(source, filename))
+        # We don't check the exit code here because it is non-zero if the source
+        # and destination are the same. The ls below will cause an error if the
+        # file was not where we expected it to be.
+        self._gem5_shell("{} cp {} {}".format(self.busybox_wrapper, source, filename),
+                        check_exit_code=False)
+        self._gem5_shell("sync")
+        self._gem5_shell("ls -la {}".format(filename))
+        logger.debug('Finished the copy in the simulator')
+        self._gem5_util("writefile {}".format(filename))
+
+        if 'cpu' not in filename:
+            while not os.path.exists(os.path.join(self.gem5out_dir, filename)):
+                time.sleep(1)
+
+        # Perform the local move
+        shutil.move(os.path.join(self.gem5out_dir, filename), dest)
+        logger.debug("Pull complete.")
+
+    def execute(self, command, timeout=1000, check_exit_code=True, background=False,
+            as_root=False, wrap_busybox=False, strip_colors=True):
+        """
+        Execute a command on the gem5 platform
+        """
+        self._check_ready()
+        if as_root and not self.is_rooted:
+            raise TargetError('Attempting to execute "{}" as root on unrooted device.'.format(command))
+        if wrap_busybox:
+            if not self.is_rooted:
+                raise TargetError('Attempting to execute "{}" with busybox. '.format(command) +
+                                  'Busybox can only be deployed to rooted devices.')
+            command = ' '.join([self.busybox_wrapper, command])
+        if background:
+            logger.debug("Attempt to execute in background. Not supported "
+                              "in gem5, hence ignored.")
```

This should raise an error to alert the user.